driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedTestSkips.java

```diff
@@ -56,69 +305,210 @@ public Skip skipJira(final String skip) {
+            return new TestApplicator(this, skip);
+        }
+
+        /**
+         * Test is skipped because the Java Reactive driver cannot comply with the spec.
+         *
+         * @param skip reason for skipping the test
+         */
+        public TestApplicator skipNoncompliantReactive(final String skip) {
+            return new TestApplicator(this, skip);
+        }
+
+        /**
+         * The test is skipped, as specified. This should be paired with a
+         * "when" clause.
+         */
+        public TestApplicator skipAccordingToSpec(final String skip) {
+            return new TestApplicator(this, skip);
+        }
+
+        /**
+         * The test is skipped for an unknown reason.
+         */
+        public TestApplicator skipTodo(final String skip) {
+            return new TestApplicator(this, skip);
+        }
+
+        public TestApplicator modify(final Modifier... modifiers) {
+            return new TestApplicator(this, Arrays.asList(modifiers));
+        }
+
+        public boolean isReactive() {
+            return reactive;
+        }
+
+        public boolean wasAssignedModifier(final Modifier modifier) {
+            return this.modifiers.contains(modifier);
         }
     }
 
-    public static final class Skip {
+    /**
+     * Applies settings to the underlying test definition. Chainable.
+     */
+    public static final class TestApplicator {
         private final TestDef testDef;
-        private final String reason;
 
-        private Skip(final TestDef testDef, final String reason) {
+        private final boolean shouldSkip;
+        @Nullable
+        private final String reasonToApply;
+        private final List<Modifier> modifiersToApply;
+        private Supplier<Boolean> shouldApply;
+        private boolean matchWasPerformed = false;
+
+        private TestApplicator(
+                final TestDef testDef,
+                final List<Modifier> modifiersToApply) {
+            this.testDef = testDef;
+            this.shouldSkip = false;
+            this.reasonToApply = null;
+            this.modifiersToApply = modifiersToApply;
+        }
+
+        private TestApplicator(
+                final TestDef testDef,
+                @NonNull
+                final String reason) {
             this.testDef = testDef;
-            this.reason = reason;
+            this.shouldSkip = true;
+            this.reasonToApply = reason;
+            this.modifiersToApply = new ArrayList<>();
+        }
+
+        private TestApplicator onMatch(final boolean match) {
+            matchWasPerformed = true;
+            if (shouldApply != null && !shouldApply.get()) {
+                return this;
+            }
+            if (shouldSkip) {
+                assumeFalse(match, reasonToApply);
+            } else {
+                if (match) {
+                    this.testDef.modifiers.addAll(this.modifiersToApply);
+                }
+            }
+            return this;
```

Comment:
For 1, the constructors are not used directly. I am not sure how to interpret this, except to produce multiple methods (`skipJira(string)` and `skipJira(string, condition)`), but then the readability suffers when the API is used.

For 2, I am not sure if the suggestion is to move invocations to a separate file (I think it is better to show how we are handling a specific set of tests in one place), or to copy/extend TestApplicator (I would need to better understand the benefits of this).

For 3, I think this focuses on the internals, rather than the API surface, which is less important. For any API, the most important thing (in my opinion) is how easy it is to read, then how easy it is to use, and only then how easy it is to modify the internals (however, I do not see these internals as particularly complex).